src/Stratis.Bitcoin.Features.BlockStore/AddressIndexing/AddressIndexer.cs

```diff
@@ -0,0 +1,291 @@
+
+        private readonly NodeSettings nodeSettings;
+
+        private readonly Network network;
+
+        private readonly IBlockStore blockStore;
+
+        private readonly INodeStats nodeStats;
+
+        private readonly ILogger logger;
+
+        private readonly DataFolder dataFolder;
+
+        private readonly IConsensusManager consensusManager;
+
+        private const string DbKey = "AddrData";
+
+        private SubscriptionToken blockConnectedSubscription, blockDisconnectedSubscription;
+
+        /// <summary>Queue that is populated when block is connected or disconnected.</summary>
+        /// <remarks><c>bool</c> key is <c>true</c> when block is connected.</remarks>
+        private AsyncQueue<KeyValuePair<bool, ChainedHeaderBlock>> blockReceivedQueue;
+
+        private LiteDatabase db;
+
+        private LiteCollection<AddressIndexerData> dataStore;
+
+        private AddressIndexerData addressesIndex;
+
+        public AddressIndexer(NodeSettings nodeSettings, ISignals signals, DataFolder dataFolder, ILoggerFactory loggerFactory,
+            Network network, IBlockStore blockStore, INodeStats nodeStats, IConsensusManager consensusManager)
+        {
+            this.signals = signals;
+            this.nodeSettings = nodeSettings;
+            this.network = network;
+            this.blockStore = blockStore;
+            this.nodeStats = nodeStats;
+            this.dataFolder = dataFolder;
+            this.consensusManager = consensusManager;
+
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+        }
+
+        public void Initialize()
+        {
+            if (this.nodeSettings.TxIndex)
+            {
+                string dbPath = Path.Combine(this.dataFolder.RootPath, "addressindex.litedb");
+                this.db = new LiteDatabase(new ConnectionString() {Filename = dbPath});
+
+                this.logger.LogDebug("TxIndexing is enabled.");
+
+                this.dataStore = this.db.GetCollection<AddressIndexerData>(DbKey);
+
+                this.addressesIndex = this.dataStore.FindAll().FirstOrDefault();
+
+                if (this.addressesIndex == null)
+                {
+                    this.logger.LogDebug("Tip was not found, initializing with genesis.");
+
+                    this.addressesIndex = new AddressIndexerData()
+                    {
+                        TipHash = this.network.GenesisHash.ToString(),
+                        AddressIndexDatas = new List<AddressIndexData>()
+                    };
+                    this.dataStore.Insert(this.addressesIndex);
+                }
+
+                this.blockReceivedQueue = new AsyncQueue<KeyValuePair<bool, ChainedHeaderBlock>>(this.OnEnqueueAsync);
+
+                // Subscribe to events.
+                this.blockConnectedSubscription = this.signals.Subscribe<BlockConnected>(blockConnectedData =>
+                {
+                    while (this.blockReceivedQueue.Count > 100)
```

Comment:
This can be prevented by fetching a block if `ChainedHeaderBlock.Block` is empty in the dequeue delegate.

Reply:
Not sure what you mean. Could you elaborate? 

I wanted to stop consensus here to prevent bloating RAM with blocks that are not processed

Reply:
If you fetch a block from disk it wont get cached, so if indexer is falling behind it can start fetching from disk and slowly catch up.

Reply:
Should it do it like that though? 

It will allow consensus sync faster but increase hard drive load. 

Also this code is almost never executed, txindexing itself is a bigger bottleneck

Reply:
ok fine with me, if we find its actually a bottle neck then we address the issue 👍